core/app/components/FormBuilder/ElementPanel/InputComponentConfigurationForm.tsx

```diff
@@ -119,7 +119,7 @@ export const InputComponentConfigurationForm = ({ index }: Props) => {
 	const { schemaName } = selectedElement;
 	const allowedComponents = componentsBySchema[schemaName];
 
-	const form = useForm<ConfigFormData>({
+	const form = useForm<ConfigFormData<(typeof allowedComponents)[number]>>({
```

Comment:
We could also solve this by making typing `ConfigFormData` like so

```ts
type ConfigFormData<I extends InputComponent = InputComponent> = /*...*/
```

However this can quickly become kind of messy as when you forget to pass in an `InputComponent` when necessary you get an incorrect type.

Alternatively, this line could be

```suggestion
	const form = useForm<ConfigFormData<InputComponent>({
```

Might be more efficient for the compiler.

Reply:
to be perfectly honest, I don't really understand what that first type means. any relevant reading material or docs you can link me to?

and for your suggested change, how are those equivalent? `(typeof allowedComponents)[number]` is more specific than `InputComponent` right?

Reply:
Sorry, that was kind of vague.

Currently `ConfigFormData` is typed like this

```ts
type ConfigFormData<I extends InputComponent> = /*...*/
```
It thus requires an `InputComponent` argument.

By instead adding the `= InputComponent`
```ts
type ConfigFormData<I extends InputComponent> = /*...*/
```

we can use it like 

```ts
const form = useForm<ConfigFormData>( // ...
```
instead, as we don't need to pass in a type argument anymore.


As for the suggested change, it is currently equivalent because `(typeof allowedComponents)[number]` is the same as `InputComponent`, as all possible `InputComponent`s are allowed.

Reply:
You may notice that this doesn't feel that strict. Ideally I would like a situation like this:

```ts

const component = InputComponent.Text

const form = useForm<ConfigFormData<typeof component>>(/*...*/)
```

But we can't really do that, because the `form` is the thing setting the `component` as well. 


So sadly we end up with a kind of "incorrect" `ConfigFormData` that looks like so

```ts
// as a reminder, this is what ConfigFormData looks like
export type ConfigFormData<I extends InputComponent> = {
	required: boolean | null;
	component: I;
	config: Static<(typeof componentConfigSchemas)[I]>;
};

declare const config: ConfigFormData<InputComponent>
// type looks like so
typeof config = {
	required: boolean | null;
	component: InputComponent.TextArea | InputComponent.Number | /*...*/;
	config: Static<(typeof componentConfigSchemas)[InputComponent.TextArea | InputComponent.Number | /*...*/]>;
}
```

This is sadly the unfortunate way that Typescript handles unions in generics: it doesn't really do any "spreading"/"iterating" per se, it just fills it in in every place you put the generic.

You can force typescript to do this spreading, by doing something like this

```ts
export type ConfigFormDataIterated<I extends InputComponent> =  I extends I ? {
	required: boolean | null;
	component: I;
	config: Static<(typeof componentConfigSchemas)[I]>;
} : never;
```
As a refresher, `X extends Y` is roughly "is X a subtype of Y". This is a conditional (the only way to do them in typescript) which is always true, so we will always go to the part after `?`. The false condition, `: never`, never happens, so we just put `never` there.

"Ok cool thomas but why the hell are we doing that?" Doing `I extends I` (or `I extends any`) here is kind of like "for each element in the union I". This is really annoying, I wish there was another way to iterate over unions in typescript, but there is not.

Now the result type will look like this

```ts
declare const config: ConfigFormDataIterated<InputComponent>
// type looks like so
typeof configIterated = 
{
	required: boolean | null;
	component: InputComponent.TextArea;
	config: Static<(typeof componentConfigSchemas)[InputComponent.TextArea>;
} | {
	required: boolean | null;
	component: InputComponent.Checkbox;
	config: Static<(typeof componentConfigSchemas)[InputComponent.Checkbox>;
} | /*...*/

```

This is a much more accurate discriminated union. Now you could theoretically do

```ts
if(configIterated.component === InputComponent.TextArea){
	config.config // <-- this would now be the correct config. Before it would just be the union of all configs.
}
```

So why not do this? I did try this, but if you want this to work properly, you also need to do it for `ComponentConfigFormProps`. When I did though, everything got realllly messy, and I couldn't figure it out anymore.

So I thought: we mostly just need this to be very accurate when defining the InputComponents themselves. Everywhere else we aren't doing anything like

```ts
if(component === InputComponent.TextArea){
  // do something with the config 
}
```

Instead we are just having each Component figure that out for itself. And at the component level we can be fully sure what the type is, since we only define the component for one component anyway.

https://github.com/pubpub/platform/blob/38afca31b59988735895e003dadcdc7bb004defe/core/app/components/FormBuilder/ElementPanel/ComponentConfig/Checkbox.tsx#L8

So I thought it best not to introduce more esoteric typescript into the codebase if we aren't going to benefit from it!